spock-core/src/main/java/org/spockframework/mock/runtime/MockMakerRegistry.java

```diff
@@ -0,0 +1,154 @@
+      }
+    } else {
+      preferredMockMakerId = null;
+    }
+    //Sort by preferredMockMakerId then by priority and if that is the same by ID
+    makerListLoc.sort(Comparator.comparing(IMockMaker::getId, (a, b) -> {
+        if (Objects.equals(a, b)) {
+          return 0;
+        }
+        //The dynamicProxy shall always win, because it uses public Java JRE API.
+        if (Objects.equals(MockMakerId.dynamicProxy, a)) {
+          return -1;
+        }
+        if (Objects.equals(MockMakerId.dynamicProxy, b)) {
+          return 1;
+        }
+        if (preferredMockMakerId != null) {
+          if (Objects.equals(preferredMockMakerId, a)) {
+            return -1;
+          }
+          if (Objects.equals(preferredMockMakerId, b)) {
+            return 1;
+          }
+        }
+        return 0;
+      })
+      .thenComparing(IMockMaker::getPriority)
+      .thenComparing(IMockMaker::getId));
+
+    this.makerList = Collections.unmodifiableList(makerListLoc);
+  }
+
+  public List<IMockMaker> getMakerList() {
+    return makerList;
+  }
+
+  public Object makeMock(IMockCreationSettings settings) throws CannotCreateMockException {
+    String mockMakerId = settings.getMockMakerId();
+    if (mockMakerId != null) {
+      IMockMaker mockMaker = makerMap.get(mockMakerId);
+      if (mockMaker == null) {
+        throw new CannotCreateMockException(settings.getMockType(), " because MockMaker with ID '" + mockMakerId + "' does not exist.");
+      }
+      return mockMaker.makeMock(settings);
+    }
+    return createWithAppropriateMockMaker(settings);
+  }
+
+  /**
+   * Returns information about a mock object, or {@code null} it the object is no mock.
+   *
+   * @param object a mock object
+   * @return information about the mock object
+   */
+  public IMockObject asMockOrNull(Object object) {
+    if (object instanceof ISpockMockObject) {
+      return ((ISpockMockObject) object).$spock_get();
+    }
+    for (IMockMaker mockMaker : makerList) {
+      IMockObject mock = mockMaker.asMockOrNull(object);
+      if (mock != null) {
+        return mock;
+      }
+    }
+    return null;
+  }
+
+  private Object createWithAppropriateMockMaker(IMockCreationSettings settings) {
+    for (IMockMaker mockMaker : makerList) {
+      if (mockMaker.isTypeMockable(settings).mockable()) {
+        return mockMaker.makeMock(settings);
+      }
+    }
+    ITypeMockability typeMockable = collectNotMockableReasons(settings);
+    throw new CannotCreateMockException(settings.getMockType(), "." + NL + typeMockable.nonMockableReason());
+  }
+
+  private ITypeMockability collectNotMockableReasons(IMockCreationSettings settings) {
+    return () -> makerList.stream()
+      .map(m -> {
+        ITypeMockability typeMockable = m.isTypeMockable(settings);
+        if (!typeMockable.mockable()) {
+          return m.getId() + ": " + typeMockable.nonMockableReason();
+        }
+        return null;
+      })
+      .filter(Objects::nonNull)
```

Comment:
If we include the id in the `ITypeMockability` we could have a default `nonMockableReasonWithId` method, and make this stream a bit nicer.
```suggestion
      .map(IMockMaker::isTypeMockable)
      .filter(m -> !m.mockable())
      .map(ITypeMockability::nonMockableReasonWithId)
```

Reply:
I would not like because then the nice way to convert a lambda to an `IMockabilityResult` would no longer work.
And this ungl code here, we need only write once, but there are multiple `IMockMaker`. which need to create an `IMockabilityResult`.